docs/architecture/adr-004-coroutine-effect-system.md

```diff
@@ -0,0 +1,303 @@
+6. For asynchronous effects (like `GetValue`), the handler resumes consensus immediately, and later provides the result as a new input.
+
+## Consequences
+
+### Positive
+
+1. **Separation of concerns**: The consensus algorithm code remains focused on the state machine logic without environment dependencies.
+2. **Code readability**: The consensus code retains a linear, procedural flow despite the need for external interactions.
+3. **Flexibility**: The same consensus core can work in different execution environments (async runtimes, actor systems, etc.)
+4. **Testability**: Effects are explicit and can be easily mocked for testing.
+5. **Error handling**: Clear points where environment errors can be handled without complicating the consensus core.
+
+### Negative
+
+1. **Learning curve**: The coroutine-based approach might be unfamiliar to some developers.
+2. **Effect handling consistency**: Requires careful documentation and examples to ensure users handle effects correctly.
+3. **Complexity with asynchronous effects**: The pattern for handling asynchronous effects like `GetValue` requires additional understanding.
+
+## Consequences
+
+### Positive
+
+1. **Separation of concerns**: The consensus algorithm code remains focused on the state machine logic without environment dependencies.
+2. **Code readability**: The consensus code retains a linear, procedural flow despite the need for external interactions.
+3. **Flexibility**: The same consensus core can work in different execution environments (async runtimes, actor systems, etc.)
+4. **Testability**: Effects are explicit and can be easily mocked for testing.
+5. **Error handling**: Clear points where environment errors can be handled without complicating the consensus core.
+
+### Negative
+
+1. **Learning curve**: The coroutine-based approach might be unfamiliar to some developers.
+2. **Debugging complexity**: Debugging code that yields and resumes can be more challenging than fully synchronous code.
+3. **Effect handling consistency**: Requires careful documentation and examples to ensure users handle effects correctly.
+4. **Complexity with asynchronous effects**: The pattern for handling asynchronous effects like `GetValue` requires additional understanding.
+
+## Implementation Notes
+
+1. The coroutine implementation relies on the `genawaiter` crate, which allows defining functions which can yield values and be resumed later.
+2. Each `Effect` variant carries a value that implements `Resumable`, which knows how to create the appropriate `Resume` variant.
+3. The `process!` macro handles the boilerplate of creating the coroutine, handling effects, and resuming execution.
+4. For asynchronous effects, the consensus must be resumed immediately, and the result must be provided later as a new input.
+5. Error handling is done at the effect handler level, with fallback behaviors defined to allow consensus to continue even if an operation fails.
```

Comment:
I'd find it helpful to include concrete code snippets here. 

Or simply point to the "Example" section below.